src/tiled/automapper.cpp

```diff
@@ -712,8 +712,18 @@ static bool compareLayerTo(const TileLayer *setLayer,
                 bool matchListYes = false;
                 bool matchListNo  = false;
 
-                if (!setLayer->contains(x + offset.x(), y + offset.y()))
-                    return false;
+
+                if (!setLayer->contains(x + offset.x(), y + offset.y())) {
+                    foreach (const TileLayer *comparedTileLayer, listYes) {
+                        if (!comparedTileLayer->contains(x, y))
+                            return false;
```

Comment:
> I understand that rather than discarding the whole match just because the output layer does not contain every position in the rule region,

right.

> in the rule region, you're instead taking a shortcut in only checking the listYes, since the listNo is irrelevant for determining an out-of-range match?

right.

The assumption I made here is 'treat any position outside the map as if it contains no tile, i.e. the nullptr tile in the input region, but it is part of the region to inspect'.

> What I don't understand yet, is why when a layer in listYes does not contain this rule region position, should the match fail? Shouldn't that be treated like the cell being empty, thus doing continue?

Looking at https://github.com/bjorn/tiled/issues/1224, the first solution I came up with in code here looked like 
![1224_wrong](https://cloud.githubusercontent.com/assets/455868/15595584/37df7dae-2375-11e6-86a4-385c9ae79803.png)

which comes from 'the stone may be just outside the map', and we don't want to have that IMHO.

So when we assume there are only nulls outside, we don't have to check any of listNo, as there is nothing to exclude (because we assume nulls).

If there is a position in listYes though, that tile is asked to be unequal to null, i.e. we want to check if there is that tile is not null. I'll look into that again if we could come up with a saner logic for that example case presented in #1224. 
 
> In general, wouldn't it make sense to make sure the layers in listYes and listNo are all the same size, 

I think it was intentional to allow arbitrary sizes. 

> and all big enough to contain the ruleRegion? Then we can drop all the contains checks on them. Or is there some reason why this is not the case?

oh. I'd have to think about that.